File: dotnet/src/altgraph_shared_app/Models/Imdb/Principal.cs
using System.Text.Json.Serialization;
using Microsoft.Azure.CosmosRepository.Attributes;

namespace altgraph_shared_app.Models.Imdb
{
  public class Principal : AbstractDocument
  {
    [JsonPropertyName("tconst")]
    public string TConst { get; set; } = string.Empty;
    [JsonPropertyName("nconst")]
    public string NConst { get; set; } = string.Empty;
    [JsonPropertyName("category")]
    public string Category { get; set; } = string.Empty;

    public Principal()
    {
    }

    public Principal(string[] lineTokens)
    {
      if (lineTokens != null)
      {
        if (lineTokens.Length > 5)
        {
          Doctype = "principal";
          TConst = lineTokens[0].Trim().ToLower();
          NConst = lineTokens[2].Trim().ToLower();
          Category = lineTokens[3].Trim().ToLower();
        }
      }
    }

    public bool IsValid()
    {
      if (TConst == null)
      {
        return false;
      }
      if (NConst == null)
      {
        return false;
      }
      if (TConst.Length < 8)
      { // tt7600874
        return false;
      }
      if (NConst.Length < 8)
      { // tt7600874
        return false;
      }
      return true;
    }

    public void ScrubValues()
    {

    }

    public bool HasNecessaryRole()
    {
      if (Category == null)
      {
        return false;
      }
      if (Category.Contains("actress"))
      {
        return true;
      }
      if (Category.Contains("actor"))
      {
        return true;
      }
      if (Category.Contains("director"))
      {
        return true;
      }
      return false;
    }
  }
}

File: java/altgraph_data_app/src/main/java/org/cjoakim/cosmos/altgraph/data/common/model/imdb/Principal.java
package org.cjoakim.cosmos.altgraph.data.common.model.imdb;

import com.fasterxml.jackson.annotation.JsonInclude;
import lombok.Data;
import lombok.EqualsAndHashCode;
import org.cjoakim.cosmos.altgraph.data.common.model.AbstractDocument;

/**
 * Instances of this class represent an IMDb Principal in a Movie.
 * <p>
 * Chris Joakim, Microsoft, November 2022
 */

@Data
@EqualsAndHashCode(callSuper = false)
@JsonInclude(JsonInclude.Include.NON_NULL)
public class Principal extends AbstractDocument {
    private String tconst;
    private String nconst;
    private String category;

    public Principal() {

        super();
    }

    public Principal(String[] lineTokens) {

        this();
        if (lineTokens != null) {
            if (lineTokens.length > 5) {
                doctype = "principal";
                tconst = lineTokens[0].strip().toLowerCase();
                nconst = lineTokens[2].strip().toLowerCase();
                category = lineTokens[3].strip().toLowerCase();
            }
        }
    }

    public boolean isValid() {

        if (tconst == null) {
            return false;
        }
        if (nconst == null) {
            return false;
        }
        if (tconst.length() < 8) { // tt7600874
            return false;
        }
        if (nconst.length() < 8) { // tt7600874
            return false;
        }
        return true;
    }

    @Override
    public void scrubValues() {

    }

    @Override
    public void intern() {

        baseIntern();

        if (tconst != null) {
            tconst = tconst.intern();
        }
        if (nconst != null) {
            nconst = nconst.intern();
        }
        if (category != null) {
            category = category.intern();
        }
    }

    public boolean hasNecessaryRole() {

        if (category == null) {
            return false;
        }
        if (category.contains("actress")) {
            return true;
        }
        if (category.contains("actor")) {
            return true;
        }
        if (category.contains("director")) {
            return true;
        }
        return false;
    }
}
